### DISCUSSION TITLE
Get all UserData methods (and functions) in a given Lua instance

### DISCUSSION THREAD

**Discussion by Danielkonge** [mlua-rs/mlua#326] [Q&A]
If I have a program that implements `UserData` for multiple different Rust objects for a given Lua instance, is there any way to get a list of all of the `UserData` object names and their methods (and functions)?

I would like to use this to generate stubs for LuaLS, and I already figured out how to get elements from `globals()` and other places, but I don't see any way to get this information for `UserData`? I thought I might be able to use `UserDataRegistry`, but that seems to only contain private fields without a way to get information out? 

If I can't get the names of all the `UserData`, is there at least a way to figure out all methods (and functions) for the `UserData` of a given name?

**Comment by khvzak** [comment]
`UserDataRegistry` is only a helper object that used to construct a metatable for `UserData` (once).
In general you can try to fetch userdata metatable and explore the `___index` and `__newindex` content. 
But sometimes it can be tricky if they point to a function than a table.

But proper solution would be adding this information to one of metatable fields.

**Reply by Danielkonge** [reply]
I am not sure I follow exactly what you are suggesting. Do you think the easiest solution for getting the methods of a given `UserData` object is to save information about which methods are implemented for that kind of object in a `UserData` field? 

And there is no method for figuring out which `UserData` objects are available in a given Lua instance?

**Reply by khvzak** [reply]
Do you want to explose userdata objects from Rust or from Lua?

> there is no method for figuring out which UserData objects are available in a given Lua instance?

What do you mean? In general there is no simple way to track how many objects of a particular type exist in Lua.

**Reply by Danielkonge** [reply]
> Do you want to explose userdata objects from Rust or from Lua?

I want to explore the `UserData` objects from Rust if possible, but I don't need information about a specific instance of the object if that makes sense. 

I would ideally want a way to know the name of all structs that implement `UserData` and then a way to figure out which methods (maybe even just the names of the methods) each of these `UserData` objects define in their implementation of `UserData`. 

For now I have thought about creating my own extra struct on the side (that would have nothing to do with Lua, but just include me writing the information I need in a separate struct) for each struct that implements `UserData`. But it would be nice if there was some easier way to get the information I need. 

> What do you mean? In general there is no simple way to track how many objects of a particular type exist in Lua. 

I am not interested in knowing how many objects of a particular type exists, instead I am interested in knowing what types exists in my current Lua environment (does that makes sense?).
